src/acl/external/session/ext_session_acl.cc

```diff
@@ -197,14 +197,11 @@ copyValue(void *dst, const DB_ENTRY *src, size_t sz)
 static int session_active(const char *details, size_t len)
 {
 #if USE_BERKLEYDB
-    DBT key = {};
+    DBT key = {(void *)details, len};
     DBT data = {};
-    key.data = (void *)details;
-    key.size = len;
 #elif USE_TRIVIALDB
-    TDB_DATA key;
-    TDB_DATA data;
-    (void)len;
+    TDB_DATA key = {(unsigned char *)details, len};
+    TDB_DATA data = {};
 #else
```

Comment:
> Done, although we have to remove the const-ness of the first argument to fetchKey

@kinkie, did you forget to push your changes to GitHub? GitHub shows no commits since my last review...

As for const-ness, that was exactly my point! Using explicit const_cast is a much better way to show that we are forced to do dangerous stuff here (compared to C-style cast that hides const removal from the casual observer). I expect to see two casts in each of the two initialization cases above -- const_cast and (static_cast or, where _necessary_, reinterpret_cast).

Reply:
Pushed and used c++ casts

Reply:
Missing casts for the `USE_BERKLEYDB` case. Since we did not add wrapper API/types that would eliminate the need to worry about these low-level concerns, then we should keep the two areas in sync manually.

Reply:
> Francesco at https://github.com/squid-cache/squid/pull/1064#issuecomment-1152944711: casts are not needed for the Berkeleysb case. ... they’re pointless

Sigh. There are only two possibilities here:

1. Berkley DB code needs no casting. This PR, which modified the corresponding code, should stop casting (i.e. remove pointless casts in the modified code).
2. Berkley DB code needs casting. This PR, which modified the corresponding code, should use C++ cast(s).

I am pretty sure the correct option is 2 because of the const_cast necessity, but, if you disagree, then you should go with option 1 and see whether the casting-free code compiles. If it does not compile, switch to option 2 (which represents the changes I [suggested](https://github.com/squid-cache/squid/pull/1064#pullrequestreview-997379150) several comments ago). Either way, this change request will be fully addressed.

Reply:
Oof. The situation is probably worse than I thought! I tried to improve this code by switching to the missing C++ casts, but a PR-based [test](https://godbolt.org/z/WoW5ceKM9) (before any of my changes) gets warnings from [clang](https://godbolt.org/z/11vG93Yea) and [gcc](https://godbolt.org/z/WoW5ceKM9):

```
<source>: In function 'int session_active(const char*, size_t)':
<source>:27:33: warning: narrowing conversion of 'len' from 'size_t' {aka 'long unsigned int'} to 'u_int32_t' {aka 'unsigned int'} [-Wnarrowing]
   27 |     DBT key = {(void *)details, len};
      |                                 ^~~
<source>:27:36: warning: missing initializer for member 'DBT::ulen' [-Wmissing-field-initializers]
   27 |     DBT key = {(void *)details, len};
      |                                    ^
<source>:27:36: warning: missing initializer for member 'DBT::dlen' [-Wmissing-field-initializers]
<source>:27:36: warning: missing initializer for member 'DBT::doff' [-Wmissing-field-initializers]
<source>:27:36: warning: missing initializer for member 'DBT::flags' [-Wmissing-field-initializers]
```

It is possible that my test is not applicable -- I have not tested Squid with Berkeley DB -- but the warnings appear to be legitimate. If I have the time, I will try to rewrite to avoid them, but I am just sharing this news for now.

Reply:
I am starting to think that the whole approach here is broken. What's wrong with having the helper support both, if the libraries can be found? This would at least improve testability.

Reply:
I pushed two commits that fix compiler warnings in examples I shared [earlier](https://github.com/squid-cache/squid/pull/1064/files#r903981342). Those two commits also address this change request. They were done under our EasierDone(ThanSaid) experiment; please feel free to revert, discard, or adjust these **completely untested** changes as needed.

> I am starting to think that the whole approach here is broken. What's wrong with having the helper support both, if the libraries can be found? This would at least improve testability.

I am not sure supporting two libraries in the same Squid build would be easier. However, I hope that I do not need to make that decision. I will leave it to those who have the time to improve this helper ;-).